src/TriggerCandidateMakerTrackFinding.cpp

```diff
@@ -0,0 +1,166 @@
+ * received with this code.
+ */
+
+#include "triggeralgs/TrackFinding/TriggerCandidateMakerTrackFinding.hpp"
+
+#include "TRACE/trace.h"
+#define TRACE_NAME "TriggerCandidateMakerTrackFindingPlugin"
+
+#include <vector>
+#include <math.h>
+using namespace triggeralgs;
+
+void
+TriggerCandidateMakerTrackFinding::operator()(const TriggerActivity& activity,
+                                                std::vector<TriggerCandidate>& output_tc)
+{
+
+  std::vector<TriggerActivity::TriggerActivityData> ta_list = { static_cast<TriggerActivity::TriggerActivityData>(
+    activity) };
+
+  // Find the offset for the very first data vs system time measure:
+  if (m_activity_count == 0) {
+    using namespace std::chrono;
+    m_initial_offset = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count() - activity.time_start*16*1e-6; 
+  }
+
+  // The first time operator is called, reset window object.
+  if (m_current_window.is_empty()) {
+    m_current_window.reset(activity);
+    m_activity_count++;
+  }
+
+  // If the difference between the current TA's start time and the start of the window
+  // is less than the specified window size, add the TA to the window.
+  else if ((activity.time_start - m_current_window.time_start) < m_window_length) {
+    m_current_window.add(activity);
+  }
+  // If it is not, move the window along.
+  else {
+    m_current_window.move(activity, m_window_length);
+  }
+
+
+  // If the addition of the current TA to the window would make it longer
+  // than the specified window length, don't add it but check whether the sum of all adc in
+  // the existing window is above the specified threshold. If it is, and we are triggering on ADC,
+  // make a TA and start a fresh window with the current TP.
+  if (m_current_window.adc_integral > m_adc_threshold && m_trigger_on_adc) {
+    TLOG() << "[TCHM] m_current_window.adc_integral " << m_current_window.adc_integral << " - m_adc_threshold " << m_adc_threshold;
+    tc_number++;
+    TriggerCandidate tc = construct_tc();
+    TLOG() << "[TCHM]     tc.time_start=" << tc.time_start << " tc.time_end=" << tc.time_end << " len(tc.inputs) " << tc.inputs.size();
+    for( const auto& ta : tc.inputs ) {
+      TLOG() << "[TCHM][TA] ta.time_start=" << ta.time_start << " ta.time_end=" << ta.time_end << " ta.adc_integral=" << ta.adc_integral;
+    }
+
+    output_tc.push_back(tc);
+    m_current_window.clear();
+  }
+
+  // If the addition of the current TA to the window would make it longer
+  // than the specified window length, don't add it but check whether the number of hit channels in
+  // the existing window is above the specified threshold. If it is, and we are triggering on channels,
+  // make a TC and start a fresh window with the current TA.
+  else if (m_current_window.n_channels_hit() > m_n_channels_threshold && m_trigger_on_n_channels) {
+    tc_number++;
+    output_tc.push_back(construct_tc());
+    m_current_window.clear();
+  }
```

Comment:
Outputting TCs requires at least one of `m_trigger_on_adc` or `m_trigger_on_n_channels` to be true. There should be a guard when both are false that raises an error.